model/PermissionProvider.php

```diff
@@ -63,29 +67,67 @@ public function getPermissions(User $user, array $resourceIds) {
         return $rights;
     }
 
-    protected function getStateObjects($stateIds) {
+    protected function getStateObjects($stateIds)
+    {
         $wfmodel = $this->getServiceManager()->get(WorkflowModel::SERVICE_ID);
         $stateCache = array();
         foreach ($stateIds as $stateId) {
             if (!empty($stateId)) {
-                $stateCache[$stateId] = $wfmodel->getState($stateId);
+                $state = $wfmodel->getState($stateId);
+                $stateWithRoles = $this->addAtomicRoles($state);
+                $stateCache[$stateId] = $stateWithRoles;
             }
         }
         return $stateCache;
     }
 
     /**
      * (non-PHPdoc)
+     *
      * @see \oat\generis\model\data\PermissionInterface::onResourceCreated()
      */
-    public function onResourceCreated(\core_kernel_classes_Resource $resource) {
+    public function onResourceCreated(\core_kernel_classes_Resource $resource)
+    {
     }
 
     /**
      * (non-PHPdoc)
+     *
      * @see \oat\generis\model\data\permission\PermissionInterface::getSupportedRights()
      */
-    public function getSupportedRights() {
+    public function getSupportedRights()
+    {
         return array('WRITE', 'READ');
     }
-}
\ No newline at end of file
+
+    private function addAtomicRoles(StateObject $state): StateObject
+    {
+        $extensionManager = common_ext_ExtensionsManager::singleton();
+        $workFlowService = $this->getServiceManager()->get(ResourceWorkflowService::SERVICE_ID);
+        $stateReadRoles = $state->getReadRoles();
+        $stateWriteRoles = $state->getWriteRoles();
+
+        foreach ($workFlowService->getOption(ResourceWorkflowService::OPTION_EXTENSIONS_WITH_ROLES) as $extensionName) {
+            $manifest = $extensionManager->getExtensionById($extensionName)->getManifest();
+            $includedRoles = $manifest->getIncludedRoles();
+
+            foreach ($includedRoles as $role => $atomicRoles) {
+                if (in_array($role, $state->getReadRoles())) {
+                    $stateReadRoles = array_merge($stateReadRoles, $atomicRoles);
+                }
+
+                if (in_array($role, $state->getWriteRoles())) {
+                    $stateWriteRoles = array_merge($stateWriteRoles, $atomicRoles);
+                }
```

Comment:
Please, use strict comparison if possible
```suggestion
                if (in_array($role, $state->getReadRoles(), true)) {
                    $stateReadRoles = array_merge($stateReadRoles, $atomicRoles);
                }

                if (in_array($role, $state->getWriteRoles(), true)) {
                    $stateWriteRoles = array_merge($stateWriteRoles, $atomicRoles);
                }
```